<a href="https://colab.research.google.com/github/YoungsunChoi914/Class2022Spring/blob/main/tensorflow_hub.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### [Tensorflow Hub](https://www.tensorflow.org/hub)

fine-tuning

### image classification (e.g. [inception_v3](https://tfhub.dev/google/imagenet/inception_v3/classification/5))

In [1]:
import tensorflow.compat.v2 as tf
import tensorflow_hub as hub

In [2]:
model = hub.KerasLayer("https://tfhub.dev/google/imagenet/inception_v3/classification/5")

In [3]:
from PIL import Image
import requests
url = 'http://images.cocodataset.org/val2017/000000039769.jpg'
image = Image.open(requests.get(url, stream=True).raw)

In [14]:
import numpy as np
image = np.array(image)

In [18]:
image[:,:,0]

array([[[0.54901963, 0.09803922, 0.21960786],
        [0.5411765 , 0.08627451, 0.22352943],
        [0.5294118 , 0.08627451, 0.16470589],
        ...,
        [0.9294118 , 0.3921569 , 0.74509805],
        [0.90196085, 0.32941177, 0.86666673],
        [0.9333334 , 0.3921569 , 0.6862745 ]]], dtype=float32)

In [23]:
# reshape into shape [batch_size, height, width, num_channels]
image = tf.reshape(image, [1, image.shape[0], image.shape[1], image.shape[2]])


InvalidArgumentError: ignored

In [19]:
# Use `convert_image_dtype` to convert to floats in the [0,1] range.
image = tf.image.convert_image_dtype(image, tf.float32)

In [22]:
image[0,:,:,0]

<tf.Tensor: shape=(480, 640), dtype=float32, numpy=
array([[0.54901963, 0.5647059 , 0.57254905, ..., 0.36862746, 0.41960788,
        0.40000004],
       [0.5411765 , 0.5568628 , 0.54509807, ..., 0.4039216 , 0.45098042,
        0.37647063],
       [0.5294118 , 0.5882353 , 0.5568628 , ..., 0.4039216 , 0.42352945,
        0.3647059 ],
       ...,
       [0.9294118 , 0.882353  , 0.92549026, ..., 0.67058825, 0.70980394,
        0.5764706 ],
       [0.90196085, 0.8862746 , 0.9333334 , ..., 0.44705886, 0.4039216 ,
        0.34901962],
       [0.9333334 , 0.96470594, 0.9333334 , ..., 0.2901961 , 0.2901961 ,
        0.28627452]], dtype=float32)>

이미지를 array 형태로 바꾸고, reshape 해야 사용 가능. 

이미지가 3차원의 array인데, 모델 특성상 4차원의 array로 바꿔야 함. 

0~1 사이 숫자로 바꿔줌

In [24]:
logits = model(image)
np.argmax(logits)

282

np.argmax 가장 높은 값 찾는 함수

282번째가 가장 큼. 

> https://storage.googleapis.com/download.tensorflow.org/data/ImageNetLabels.txt